[turner.lee] 지하철 노선도 미션 리뷰 요청드립니다.

sumniy opened the pull request:
안녕하세요 리뷰어님, 터너입니다.
아직 3단계의 요구사항을 Section 등록 기능에 적용하지는 못했지만, 주말전까지 리뷰를 보내는게 좋다고 하셔서 리뷰 요청드리게 되었습니다.
리뷰 잘 부탁드립니다. 감사합니다.

sumniy committed:
Feat: StationController 구현

- StationDao
- createStation
- showStations
  - stationDao의 findAll 메서드를 통해 조회 후 반환하도록 구현
- deleteStation
  - stationDao의 deleteById를 사용하도록 변경

sumniy committed:
Feat: 노선 관리 기능 구현

기능 목록
- 노선 생성, 중복 노선 생성 불가
- 개별 노선 조회, 노선 목록 조회
- 노선 삭제

sumniy committed:
Refactor: 코드 리팩토링

조건문 분리하여 명확한 의미를 알 수 있도록 함.

sumniy committed:
Refactor: stationDao static으로 변경

StationController외의 클래스에서 활용할 수 있도록
static으로 변경

sumniy committed:
Feat: Section 관련 기능 구현

Line에 Section 추가, 삭제, 수정 등의 기능을 구현

sumniy committed:
Etc: package.json 업데이트

sumniy committed:
feat : stationDao jdbc 적용

jh280722 committed:
feat : LineDao jdbc 적용

jh280722 committed:
feat : lineDao jdbc 적용

jh280722 committed:
feat : SectionDao jdbc 변환 중

jh280722 committed:
feat : SectionDao jdbc 변환 중

findstar commented on src/main/java/subway/line/Line.java:
```diff
@@ -0,0 +1,44 @@
+
+public class Line {
+    private static final int FIRST_INDEX = 0;
+    public static final int MIN_SECTION_SIZE = 1;
```
이 상수값은 사용하는 곳이 없습니다.

findstar commented on src/main/java/subway/station/StationDao.java:
```diff
@@ -1,33 +1,58 @@
     public Station save(Station station) {
-        Station persistStation = createNewObject(station);
-        stations.add(persistStation);
+        String sql = "insert into station (name) values (?)";
```
동일한 이름의 역이 추가되면 어떻게 되나요?

sumniy replied:
현재는 500에러가 발생하네요. 예외 처리하겠습니다. 감사합니다!

findstar commented on src/main/java/subway/station/StationDao.java:
```diff
@@ -1,33 +1,58 @@
-        field.setAccessible(true);
-        ReflectionUtils.setField(field, station, ++seq);
-        return station;
+    public Station findById(Long id) {
```
id 에 해당하는 역을 찾을 수 없는 경우는 어떻게 되나요?

sumniy replied:
이 부분도 예외처리 하겠습니다. 😀

findstar commented on src/main/java/subway/line/LineController.java:
```diff
@@ -0,0 +1,108 @@
+
+    @PutMapping("/lines/{id}")
+    public ResponseEntity updateLine(@RequestBody LineRequest lineRequest, @PathVariable Long id) {
+        Line line = lineDao.findById(id);
+        Line updateLine = new Line(lineRequest.getName(), lineRequest.getColor());
```
새로운 Line 객체를 생성하기 보다는 새로운 값을 셋팅하는 방법으로 구현하면어떨까요?
LineDao 의 update 메서드에는 수정된 Line 객체를 전달하도록 변경하면 됩니다. 

차후 ORM 을 사용하는 코드에서는 쿼리가 아닌 객체 자체를 다루게 되는데요. 
이를 대비해서 객체 자체를 다루는 방법으로 풀어보면 좋겠습니다. 
물론 이를 위해서 Line 객체에 Setter 를 생성하는 방식보다는 이름과 컬러를 변경하는 메서드를 추가하여 
기능을 제공하면 됩니다.

sumniy replied:
넵 수정하는 방식으로 반영하겠습니다.

findstar commented on src/main/java/subway/line/LineResponse.java:
```diff
@@ -11,6 +11,15 @@
     private int extraFare;
     private List<StationResponse> stations;
 
+    public LineResponse() {
```
사용하지 않는 생성자 입니다.

findstar commented on src/main/java/subway/line/SectionDao.java:
```diff
@@ -0,0 +1,135 @@
+
+        KeyHolder keyHolder = new GeneratedKeyHolder();
+        jdbcTemplate.update(connection -> {
+            PreparedStatement ps = connection.prepareStatement(sql, new String[] { "id" });
```
가급적 줄여서 작성하지 않는 가이드가 있는데요, 
일반적으로 PreparedStatement 는 `pstmt` 라고 줄여서 표현하기도 합니다.

findstar replied:
Statement.RETURN_GENERATED_KEYS 를 사용할 수 있습니다.

findstar commented on src/main/java/subway/line/SectionDao.java:
```diff
@@ -0,0 +1,135 @@
+        return persistSection;
+    }
+
+    public void add(Section newSection) {
+        checkSameSection(newSection);
+        checkNoStation(newSection);
+        checkOneSection();
+    }
```
save 와 add 의 역할은 무엇이 다른가요? Dao 에서 add 메서드의 역할은 무엇일까요?

sumniy replied:
save는 db 저장, add는 새로운 section을 추가하는 메서드인데 dao가 section에 대한 로직을 담당하다보니 모호해진 표현인 것 같습니다. 
add 메서드를 sections로 분리해보겠습니다. 피드백 감사합니다!

findstar reviewed (changes requested):
안녕하세요, 터너. 피드백 진행하게된 제임스 입니다. 
아직 미션 중인점 감안하고 리뷰 진행했습니다. 

먼저 예외상황에 대한 고려가 추가되면 좋겠습니다. 
대표적으로 중복된 역의 이름을 추가하는 경우와 같이 처리할 수 없는
요청에 대한 예외처리가 필요합니다. 

SectionDao 에서 새로운 Section 을 추가하기 위해서 
복잡한 로직들을 수행하는데요. Sections 일급 콜렉션을 만들어서 
해당 Line 의 Sections 에 대한 조회및 중복여부를 확인할 수 있습니다. 

코멘트 참고부탁드리며 
잘 이해가되지 않는 부분은 언제든지 DM 부탁드립니다.
그럼 계속 화이팅입니다! 😄

dsyun96 committed:
feat : 구간 추가 기능 구현

dsyun96 committed:
refactor : 전체 코드 리팩터링

sumniy committed:
Refactor: Station 관련 예외 처리 추가

- 쿼리를 실행하는 메서드가 어떤 Exception을 발생시킬 수 있는 지 명시
- deleteById는 실행결과를 확인하여 예외 발생하도록 변경
- StationController에서는 ExceptionHandler를 추가하여 발생하는 Exception들을 관리.

sumniy committed:
Refactor: Line 관련 예외 처리 추가

findByName으로 쿼리를 실행해 중복을 확인하지 않고,
insert할 때 DuplicateKeyException이 발생하는지 여부로
중복된 이름의 Line 삽입에 대한 예외를 처리함.

- 쿼리를 실행하는 메서드가 어떤 Exception을 발생시킬 수 있는 지 명시
- deleteById는 실행결과를 확인하여 예외 발생하도록 변경
- LineController에서는 ExceptionHandler를 추가하여 발생하는 Exception들을 관리.

sumniy committed:
Refactor: Section Controller로 분리 및 관련 예외 처리 추가

Section의 생성, 삭제를 담당하는 기능을 LineController에서
SectionController로 분리함.

- deleteById는 실행결과를 확인하여 예외 발생하도록 변경
- SectionController에서는 ExceptionHandler를 추가하여 발생하는 Exception들을 관리.

sumniy commented:
안녕하세요, 제임스. 작성해주신 피드백 반영했습니다. 

피드백 관련해서
```
private LineResponse() {
    }
```
위의 생성자는 코드에서는 사용하는 부분이 없지만 jackson에서 databind 할 때 필요해서 삭제하지 않고 그대로 두었습니다.

그리고 Section만 Service가 분리되어 있는데, 이 부분은 페어프로그래밍에서 미리 진행이 되어서 그랬습니다.
도메인 분리와 Service 분리 등의 리팩토링은 다음 단계에서 반영해서 PR 하겠습니다.

sumniy commented:
진행하던 중에 궁금한 점이 몇 가지 생겨서 따로 코멘트로 질문 드립니다. 
1. 예외처리를 할 때, dao, service, domain 등 모든 부분에서 예외 처리를 해주어야 하는지 궁금합니다. 예를 들어서 dao의 save 메서드를 실행하기 전에 이미 domain에서 save를 위해 필요한 예외 처리를 했다고 하더라도 dao에서도 한 번 더 예외처리를 해줘야 하는지가 궁금합니다.

2. 예외처리의 방식에 대해서 고민이 있습니다. 
   * 현재 작성한 코드에서 중복된 이름의 station을 추가하는 예외 상황을 처리할 때에 쿼리 실행 시 발생하는 `DuplicateKeyException`을 `StationController`에서 `ExceptionHandler`를 통해 처리하게 했습니다. 
   *  다른 방법으로는 삽입 전에 select query를 실행해서 존재 여부를 if 문으로 판단해서 직접 Exception을 throw하는 방법을 생각했습니다.   

전자는 추가적인 DB접근과 확인 로직이 없다는 점이 장점이고, 후자는 코드만으로 이름에 대해 중복체크를 한다는 것을 직관적으로 알 수 있고 Exception을 직접 다루는데 장점이라고 생각했습니다.

이에 대해서 제임스는 어떻게 생각하시는지 궁금합니다. 그리고 실무에서는 예외처리를 어느정도 범위로 어떻게 하고 있는지도 궁급합니다. 

긴 질문 읽어주셔서 감사합니다.

findstar commented on src/main/java/subway/section/Section.java:
```diff
@@ -0,0 +1,83 @@
+        return new Section(lineId, lineRequest.getUpStationId(), lineRequest.getDownStationId(), lineRequest.getDistance());
+    }
+
+    public static Section of(Long lineId, SectionRequest sectionRequest) {
```
도메인 객체에서 Request 에 의존하고 있습니다. 다른 방법은 없을지 고민해봐주세요 😄

findstar commented on src/main/java/subway/section/SectionService.java:
```diff
@@ -0,0 +1,189 @@
+    private static final int SECOND_INDEX = 1;
+
+    private SectionDao sectionDao;
+    private List<Section> sections;
```
이전 미션들에서 계속 학습했던 일급 콜렉션 객체를 사용해보시면 어떨까요?

findstar reviewed (approved):
안녕하세요 제임스입니다. 
지난번 피드백 드린 부분이 개선되었네요. 수고하셨습니다. 

질문주신 부분에 답변드리면, 예외처리는 각각의 레이어에서 필요한 부분이 있습니다.
레이어를 분리한다는건 관심사를 분리한다는 의미가 되므로 각각의 레이어에서 역할에 맞는 
예외처리가 필요합니다. 
역 이름 중복의 경우에는 현재의 과제는 단순한 요구사항이기 때문에 도메인 로직에서 처리하는 것이 더 낫습니다만, 나중에 DB 를 처리할 때는 중복키에 대한 고민도 필요합니다 😄 

추가로 테스트 케이스가 일부 실패하고 있습니다. 보완 부탁드립니다. 
예외처리도 좀 더 세분화해서 진행해보세요. 
이미 4단계를 일부 진행중이시라 이 PR 은 머지하겠습니다. 
코멘트 남긴 부분 참고해서 다음 단계를 진행해주세요.
계속 화이팅입니다. 궁금하신점이 있다면 언제든지 DM 주세요 .
